src/Status.hpp

```diff
@@ -0,0 +1,155 @@
+  configuration_error = -1,
+  consumer_error = -2,
+  metadata_error = -3,
+  topic_partition_error = -4,
+  assign_error = -5,
+  topic_error = -6,
+  offset_error = -7,
+  start_time_error = -8,
+  message_error = -9,
+  write_error = -10,
+  not_initialized = -1000,
+};
+
+const std::string Err2Str(const FileWriter::StreamMasterError &);
+const std::string Err2Str(const FileWriter::StreamerError &);
+
+// Data type for collecting informations about the streamer
+class StreamerStatusType {
+  friend class StdIOWriter;
+  friend class JSONWriter;
+  friend class FlatbuffersWriter;
+
+public:
+  StreamerStatusType();
+  StreamerStatusType(const StreamerStatusType &other);
+  StreamerStatusType(StreamerStatusType &&other) noexcept = default;
+
+  ~StreamerStatusType() = default;
+
+  StreamerStatusType &operator=(const StreamerStatusType &other);
+  StreamerStatusType &operator=(StreamerStatusType &&other);
+  StreamerStatusType &operator+=(const StreamerStatusType &other);
+  StreamerStatusType &operator-=(const StreamerStatusType &other);
+  StreamerStatusType operator+(const StreamerStatusType &other);
+
+  void reset();
+
+  double bytes{0.0};
+  double messages{0.0};
+  double errors{0.0};
+  double bytes2{0.0};
+  double messages2{0.0};
+};
+
+// Data type for extracting statistics about the streamer
+struct StreamerStatisticsType {
+  double size_avg{0}, size_std{0};
+  double freq_avg{0}, freq_std{0};
+  int partitions{0};
+};
+
+class StreamMasterStatus {
+  friend class StdIOWriter;
+  friend class JSONWriter;
+  friend class FlatbuffersWriter;
+
+public:
+  StreamMasterStatus() = default;
+  StreamMasterStatus(const int &value) : status(value){};
+
+  StreamMasterStatus &push(const std::string topic_name,
+                           const StreamerStatusType &status,
+                           const StreamerStatisticsType &stats) {
+    topic.push_back(topic_name);
+    streamer_status.push_back(status);
+    streamer_stats.push_back(stats);
+    return *this;
+  }
+
+private:
+  std::vector<std::string> topic;
+  std::vector<StreamerStatusType> streamer_status;
+  std::vector<StreamerStatisticsType> streamer_stats;
+  int status{0};
+};
+
+// Collects informations about the StreamMaster and the Streamers,
+// extract statistics
+class StreamerStatus {
+  using StreamerError = FileWriter::StreamerError;
+
+  friend class StdIOWriter;
+  friend class JSONWriter;
+  friend class FlatbuffersWriter;
+
+public:
+  StreamerStatus()
+      : last_time{std::chrono::system_clock::now()}, partitions{0},
+        run_status_(StreamerError(StreamerErrorCode::stopped)) {}
+  StreamerStatus(const StreamerStatus &other)
+      : current{other.current}, last{other.last}, last_time{other.last_time},
+        partitions{other.partitions}, run_status_(other.run_status_) {}
+
+  void add_message(const double &bytes);
+  void error() { current.errors++; }
+  void error(const StreamerError &value) {
+    run_status_ = value;
+    error();
+    return;
```

Comment:
I had this thing at some moment to mark functions taking no argument with foo(void) and end functions returning void with return. My bad, I will get rid of all of them.